independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/OpenContainerPathTree.java

```diff
@@ -0,0 +1,189 @@
+
+    private static final boolean USE_WINDOWS_ABSOLUTE_PATH_PATTERN = !FileSystems.getDefault().getSeparator().equals("/");
+
+    private static volatile Pattern windowsAbsolutePathPattern;
+
+    private static Pattern windowsAbsolutePathPattern() {
+        return windowsAbsolutePathPattern == null ? windowsAbsolutePathPattern = Pattern.compile("[a-zA-Z]:\\\\.*")
+                : windowsAbsolutePathPattern;
+    }
+
+    static boolean isAbsolutePath(String path) {
+        return path != null && !path.isEmpty()
+                && (path.charAt(0) == '/' // we want to check for '/' on every OS
+                        || USE_WINDOWS_ABSOLUTE_PATH_PATTERN
+                                && (windowsAbsolutePathPattern().matcher(path).matches())
+                        || path.startsWith(FileSystems.getDefault().getSeparator()));
+    }
+
+    static void ensureResourcePath(FileSystem fs, String path) {
+        if (isAbsolutePath(path)) {
+            throw new IllegalArgumentException("Expected a path relative to the root of the path tree but got " + path);
+        }
+        // this is to disallow reading outside the path tree root
+        if (path != null && path.contains("..")) {
+            for (Path pathElement : fs.getPath(path)) {
+                if (pathElement.toString().equals("..")) {
+                    throw new IllegalArgumentException("'..' cannot be used in resource paths, but got " + path);
+                }
+            }
+        }
+    }
+
+    protected PathFilter pathFilter;
+
+    /**
+     * For deserialization of DirectoryPathTree
+     */
+    public OpenContainerPathTree() {
+    }
+
+    protected OpenContainerPathTree(PathFilter pathFilter) {
+        this(pathFilter, false);
+    }
+
+    protected OpenContainerPathTree(PathFilter pathFilter, boolean manifestEnabled) {
+        super(manifestEnabled);
+        this.pathFilter = pathFilter;
+    }
+
+    protected OpenContainerPathTree(PathFilter pathFilter, PathTreeWithManifest pathTreeWithManifest) {
+        super(pathTreeWithManifest);
+        this.pathFilter = pathFilter;
+    }
+
+    /**
+     * This is the path to the container.
+     * <p>
+     * In the case of a zip archive, it's the path of the archive.
+     * In the case of a directory, it's the directory.
+     * <p>
+     * Should only be used for equals/hashCode.
+     */
+    protected abstract Path getContainerPath();
+
+    /**
+     * This is the path to the container.
+     * <p>
+     * In the case of a zip archive, it's the path to the root of the archive (i.e. a ZipPath).
+     * In the case of a directory, it's the directory.
+     * <p>
+     * Should be used for any read operation on the container.
+     */
+    protected abstract Path getRootPath();
+
+    @Override
+    public OpenPathTree open() {
+        return this;
+    }
+
+    @Override
+    public Collection<Path> getRoots() {
+        return List.of(getRootPath());
```

Comment:
In the current implementation, it would be getContainerPath()

Reply:
Are you sure? Because in the case of an archive, we have:

- `containerPath`: the path to the archive itself (for instance the directory or the path to the jar) - this doesn't exist in the previous implementation
- `rootPath`: the path to the root of the archive. In the case of a directory, it's the same (the directory itself), in the case of an archive it's a `ZipPath` to the `/` of the archive

For me `getRoots()` returned actionable roots so they should use `getRootPath()`.

Reply:
> For me getRoots() returned actionable roots so they should use getRootPath().

For a directory-backed PathTree - yes. For an ArchivePathTree it's the archive file. But I see your point if it's meant to replace the DirectoryPathTree.

Reply:
OK, I checked the callers and apparently it's used to get the path of the archive. I made the change and added a note to the javadoc.
Will push shortly.

Reply:
There was actually one caller going through the roots as reported in: #42571 .

Reverted back to using `getRootPath()` in https://github.com/quarkusio/quarkus/pull/42588 .